### PR TITLE
I2c nak detection workarround

### PR DESCRIPTION
**Pull request by engdoreis**
This PR documents and implements the NAK detection workaround discussed in the issue #18764. 

Fix https://github.com/lowRISC/opentitan/issues/18764

### CODE REVIEW

File: hw/ip/i2c/doc/programmers_guide.md [hunk 1]
```diff
@@ -101,6 +101,30 @@ All other parameters in registers `TIMING2`, `TIMING3`, `TIMING4` are unchanged
...
+1. Address the device for writing by writing:
+   - FDATA.START = 1;
+   - FDATA.FBYTE = <7-bit address + write bit>.
+2. Fill the TX_FIFO by writing FDATA.FBYTE `n`-1 times.
+3. Send last byte with the stop bit by:
+    - FDATA.STOP = 1;
+    - FDATA.FBYTE = <last byte>.
+
+## Reading `n` bytes from a device:
+1. Address the device for reading by writing:
```
Comment (a-will) [inline hw/ip/i2c/doc/programmers_guide.md:114]: Maybe add a TODO to come back to this after we've fixed the missing behavior?
```suggestion
<!-- TODO: Fix #18917 and remove the FMT_FIFO empty check before adding the full transaction to the FMT_FIFO -->
1. Address the device for reading by writing:
```

File: hw/ip/i2c/doc/programmers_guide.md [hunk 2]
```diff
@@ -101,6 +101,30 @@ All other parameters in registers `TIMING2`, `TIMING3`, `TIMING4` are unchanged
...
+   - FDATA.START = 1;
+   - FDATA.FBYTE = <7-bit address + read bit>.
+2. Wait the write transaction to finish by checking:
+    - STATUS.FMTEMPTY == 1.
+3. If INTR_STATE.nak is 1, then go back to step 1, else proceed.
+4. Issue a read transaction by writing.
+    - FDATA.READ = 1;
+    - FDATA.STOP = 1;
+    - FDATA.FBYTE = <`n`>.
+5. Wait the read transaction to finish by checking:
```
Comment (a-will) [inline hw/ip/i2c/doc/programmers_guide.md:124]: nit:
```suggestion
5. Wait for the read transaction to finish by checking:
```

File: sw/device/lib/testing/i2c_testutils.c [hunk 3]
```diff
@@ -317,3 +314,11 @@ status_t i2c_testutils_wait_host_idle(const dif_i2c_t *i2c) {
   } while (!status.host_idle);
   return OK_STATUS();
 }
+
+status_t i2c_testutils_wait_transaction_finish(const dif_i2c_t *i2c) {
+  dif_i2c_status_t status;
+  do {
+    TRY(dif_i2c_get_status(i2c, &status));
+  } while (!status.fmt_fifo_empty);
```
Comment (a-will) [inline sw/device/lib/testing/i2c_testutils.c:322]: Small tweak (but maybe could be made in another PR):

Can we make use of the FMT_FIFO threshold and interrupt instead? This would represent the more typical pathway in practice, I think. We would still poll here, but it'd be the interrupt bit instead.

**Review by a-will — APPROVED** [review]

File: hw/ip/i2c/doc/programmers_guide.md (hunk 1 above)
Comment (engdoreis) [reply hw/ip/i2c/doc/programmers_guide.md]: Done
